docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+# Swift and C++ Interoperability Roadmap
+
+* Introduction
+    * C++, as a language
+    * C and Objective-C interoperability
+* Specifics and trade-offs
+* Goals
+* The approach 
+    * Foreign reference types
+    * Owned types and value lifetimes
+    * Iterators and ranges (by Egor)
+    * Generic APIs https://forums.swift.org/t/bridging-c-templates-with-interop/55003
+* Evolution and process
+    * Example: virtual/overridden APIs
+* How interop fits into the ecosystem more generally (lldb, ide, spm)
+    * *The* Standard Library overlay (by Egor)
+
+## Introduction
+
+**This roadmap is** a sketch, rather than a final design for the “forward” (using C++ APIs from Swift) half of C++ and Swift interoperability. This roadmap outlines some high-level topics related to C++ interoperability, overarching goals that drive the project’s design decisions, a process for evolving C++ interoperability over time, and a collection of specific problems and their potential solutions. 
+
+“Reverse” interoperability (using Swift APIs from C++) is another extremely important part of the interoperability story, however, reverse interoperability has largely different goals and constraints, which necessarily mean a different design and roadmap. Alex Lorenz will be putting together a roadmap for reverse interoperability which outlines these goals and will shed light onto the differences between the two halves of the projects. In the future this roadmap will need to be updated to explain how these two halves fit together and present the “bi-directional” interoperability story that is C++ and Swift interoperability as a whole. 
+
+**C++, as a language:** C++ is an stubbornly un-opinionated language. There is little to no consensus on anything from naming, to generic programming, value categories, or error handling (to name a few things). This lack of opinion is in sharp contrast to Swift where there is generally one, specific way to do things. Even in the C++ Standard Library there is a rather un-standard set of API models. The un-opinionated nature of C++ means there are few, if any guarantees about a given C++ API, because the same functionality may be implemented using many different language features or programming paradigms. When consuming an arbitrary C++ API, you must be prepared for *anything*. When consuming Swift APIs, this is not the case. Swift is an opinionated language which means rather than consuming *any* API pattern you are consuming *some*, specific API pattern. Swift clearly defines patterns for naming, generic programming, value categories, error handling, etc. and it is rare that a codebase deviates from the standard practice. 
```

Comment:
Yes, maybe I should make this a bit less harsh on C++. The important piece here is we're going from an unopinionated language to an opinionated one. That necessitates a certain approach for importing APIs and helps shape our goals, so I think it's important to call out.

Reply:
Instead of "opinionated", perhaps  "flexible" (or permissive) might be a better word here.

Reply:
I think perhaps you should restructure this paragraph and the next into a combined section that makes these three points in this order:

1. Swift is an opinionated language where there is a strong culture of following certain idioms.
2. Objective-C is also pretty opinionated and has a culture of following its own idioms. These idioms are different from Swift’s, but the compiler can mechanically map the Objective-C idioms to Swift idioms to present them in a way that feels natural.
3. C++ is *not* opinionated and does *not* have a culture of following certain idioms, so C++ interop will have to be different from ObjC interop. It will probably be more complicated, with more manual knobs and controls, and the results may not be as seamless to clients.

That organization would help to emphasize that you’re worrying about how to handle the mismatch between the styles of the two languages, rather than criticizing the style of C++.

Reply:
While some of what you say here is true, it seems a little incongruous at this point in my reading.  Perhaps its contribution will become clear as I go on, but my first instinct about this paragraph is that it's not serving any purpose, and since it's controversial, may be better omitted.